### PR TITLE
ZCS-7986: Create a mechanism to upgrade LDAP schema in patch release

### PR DESCRIPTION
**Pull request by nileshpawar1**
Currently, upgrades to LDAP happen only during major releases.

Solution: 
- Created a build script for generating LDAP configuration files
- Created a Perl script to use those files and update the LDAP.

### CODE REVIEW

**Review by ajinkya81 — APPROVED** [review]

File: LdapPatch/conf/attrs/amavisd-new-attrs.xml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!-- NOTE: we do not generate schema for these attrs.  This file is
+     only for the runtime use of the server and tools for verifying
+     attr values. -->
+
+<attrs>
+
```
Comment (swapnilpingle89) [inline LdapPatch/conf/attrs/amavisd-new-attrs.xml]: Please remove these empty lines

File: LdapPatch/conf/attrs/zimbra-ocs.xml [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+</objectclass>
+
+<objectclass id="13" name="zimbraObjectEntry" type="STRUCTURAL">
+  <sup>top</sup>
+  <desc>Object type info</desc>
+  <comment>
+    zimbraObjectEntry is used to represent status about object types
+  </comment>
+</objectclass>
+
```
Comment (swapnilpingle89) [inline LdapPatch/conf/attrs/zimbra-ocs.xml]: Please remove these empty lines

File: LdapPatch/upgradeLdap.pl [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+our $SU = "su - zimbra -c ";
+
+my $pwuid = getpwuid( $< );
+if ($pwuid eq "zimbra") {
+  $SU = "";
+}
+
+our %options = ();
+
+if(isInstalled("zimbra-ldap")) {
```
Comment (dasiyogesh) [inline LdapPatch/upgradeLdap.pl]: add space between if & (

File: LdapPatch/upgradeLdap.pl [hunk 4]
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/perl
+
+use strict;
+
+use Cwd;
+use Time::localtime qw(ctime);
+
+my $newinstall  = 0;
```
Comment (dasiyogesh) [inline LdapPatch/upgradeLdap.pl]: remove extra space before =

File: LdapPatch/upgradeLdap.pl [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+if(isInstalled("zimbra-ldap")) {
+  installLdapSchema();
+}
+
+sub installLdapSchema {
+  main::runAsZimbra("/opt/zimbra/libexec/zmldapschema 2>/dev/null");
+}
+
+if (!$newinstall) {
```
Comment (dasiyogesh) [inline LdapPatch/upgradeLdap.pl]: is this check required, as $newinstall is defaulted to 0 at line 8 Or is it getting updated somewhere?

Comment (nileshpawar1) [reply LdapPatch/upgradeLdap.pl]: done!

File: LdapPatch/upgradeLdap.pl [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  my $rc = 0xffff & system ("$pkgQuery > /dev/null 2>&1");
+  $rc >>= 8;
+  if (($platform =~ /^DEBIAN/ || $platform =~ /^UBUNTU/) && $rc == 0 ) {
+    $good = 1;
+    $pkgQuery = "dpkg -s $pkg | egrep '^Status: ' | grep 'not-installed'";
+    $rc = 0xffff & system ("$pkgQuery > /dev/null 2>&1");
+    $rc >>= 8;
+    return ($rc == $good);
+  } else {
```
Comment (dasiyogesh) [inline LdapPatch/upgradeLdap.pl]: else is not required, can just return ($src == good) after if block

Comment (nileshpawar1) [reply LdapPatch/upgradeLdap.pl]: done

**Review by dasiyogesh — APPROVED** [review]

**Pull request merged**